helper/schema/resource_data.go

```diff
@@ -604,6 +607,67 @@ func (d *ResourceData) GetRawConfig() cty.Value {
 	return cty.NullVal(schemaMap(d.schema).CoreConfigSchema().ImpliedType())
 }
 
+// GetRawConfigAt is a helper method for retrieving specific values
+// from the RawConfig returned from GetRawConfig. It returns the cty.Value
+// for a given cty.Path or an error diagnostic if the value at the given path does not exist.
+//
+// GetRawConfigAt is considered advanced functionality, and
+// familiarity with the Terraform protocol is suggested when using it.
+func (d *ResourceData) GetRawConfigAt(valPath cty.Path) (cty.Value, diag.Diagnostics) {
+	rawConfig := d.GetRawConfig()
+	configVal := cty.NullVal(cty.EmptyObject)
+
+	if rawConfig.IsNull() {
+		return configVal, diag.Diagnostics{
+			{
+				Severity: diag.Error,
+				Summary:  "Invalid config path",
+				Detail: "The Terraform Provider unexpectedly provided a path that does not match the current schema. " +
+					"This can happen if the path does not correctly follow the schema in structure or types. " +
+					"Please report this to the provider developers. \n\n" +
+					"The RawConfig is empty.",
+				AttributePath: valPath,
+			},
+		}
+	}
+	err := cty.Walk(rawConfig, func(path cty.Path, value cty.Value) (bool, error) {
+		if path.Equals(valPath) {
+			configVal = value
+			return false, nil
+		}
+		return true, nil
+	})
+	if err != nil {
+		return configVal, diag.Diagnostics{
+			{
+				Severity: diag.Error,
+				Summary:  "Invalid config path",
+				Detail: "The Terraform Provider unexpectedly provided a path that does not match the current schema. " +
+					"This can happen if the path does not correctly follow the schema in structure or types. " +
+					"Please report this to the provider developers. \n\n" +
+					fmt.Sprintf("Encountered error while retrieving config value %s", err.Error()),
+				AttributePath: valPath,
+			},
+		}
+	}
+
+	if configVal.IsNull() {
+		return configVal, diag.Diagnostics{
+			{
+				Severity: diag.Error,
+				Summary:  "Invalid config path",
+				Detail: "The Terraform Provider unexpectedly provided a path that does not match the current schema. " +
+					"This can happen if the path does not correctly follow the schema in structure or types. " +
+					"Please report this to the provider developers. \n\n" +
+					"Cannot find config value for given path.",
+				AttributePath: valPath,
+			},
+		}
+	}
```

Comment:
(adding comments as I run into stuff)

I don't think we should return an error in this scenario. This would also trigger if you had an optional attribute that does exist in the schema, the value is found during walk, but the actual value is null.

We could consider adding a boolean to determine if we found something during the walk.